contrib/long_distance_matching/circular_buffer_table.c

```diff
@@ -0,0 +1,107 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "ldm_hashtable.h"
+#include "mem.h"
+
+//TODO: move def somewhere else.
+//TODO: memory usage is currently no longer LDM_MEMORY_USAGE.
+//      refactor code to scale the number of elements appropriately.
+
+// Number of elements per hash bucket.
+#define HASH_BUCKET_SIZE_LOG 0 // MAX is 4 for now
+#define HASH_BUCKET_SIZE (1 << (HASH_BUCKET_SIZE_LOG))
+
+struct LDM_hashTable {
+  U32 size;
+  LDM_hashEntry *entries;  // 1-D array for now.
+
+  // Position corresponding to offset=0 in LDM_hashEntry.
+  const BYTE *offsetBase;
+  BYTE *bucketOffsets;     // Pointer to current insert position.
+
+                           // Last insert was at bucketOffsets - 1?
+};
+
+LDM_hashTable *HASH_createTable(U32 size, const BYTE *offsetBase) {
+  LDM_hashTable *table = malloc(sizeof(LDM_hashTable));
+  table->size = size;
+  table->entries = calloc(size * HASH_BUCKET_SIZE, sizeof(LDM_hashEntry));
+  table->bucketOffsets = calloc(size, sizeof(BYTE));
+  table->offsetBase = offsetBase;
+  return table;
+}
+
+static LDM_hashEntry *getBucket(const LDM_hashTable *table, const hash_t hash) {
+  return table->entries + (hash << HASH_BUCKET_SIZE_LOG);
+}
+
+LDM_hashEntry *HASH_getValidEntry(const LDM_hashTable *table,
```

Comment:
I did some experimenting with this. There is a large (> 2x?) speed hit in checking the backwards/reverse match every time (even when comparing checksums first). I approximated by checking backwards only if the forward match is long enough.  

To give a brief idea of what the speeds look like in terms of bucket size and checking all the entries (this is from 4 versions of clang--3.9.0, 3.9.1, 4.0.0, 4.0.1--tarred together):

Bucket size, speed, check first or check all?
8, 106 MB/s, first only
8, 91 MB/s, all
16, 74 MB/s, first only
16, 65 MB/s, all

I don't really notice an improvement in compression ratio (and even see it worsening slightly but that might be file dependent).